core/src/main/java/com/yahoo/oak/InternalOakHash.java

```diff
@@ -406,37 +406,35 @@ void printSummaryDebug() {
 
     private static final class IteratorState<K, V> extends BasicIteratorState<K, V> {
 
-        // lastKeyAccessed is used to find the index of the last chunk accessed
-        // but calculating the key hash and extracting the MSBs.
-        //keyBufferValid is used to check the validity of lastKeyAccessed field
-        private KeyBuffer lastKeyAccessed = null;
+        // lastKeyHash is used to find the index of the last chunk accessed by extracting the MSBs.
+        // keyBufferValid is used to check the validity of lastKeyHash field
+        private int lastKeyHash;
         private boolean keyBufferValid = false;
 
         private IteratorState(HashChunk<K, V> nextHashChunk, HashChunk<K, V>.HashChunkIter nextChunkIter,
                               int nextIndex) {
             super(nextHashChunk, nextChunkIter, nextIndex);
         }
 
-        private KeyBuffer getKeyBuffer() {
-            if (keyBufferValid) {
-                return lastKeyAccessed;
-            } else {
-                return null;
-            }
+        private void setLastKeyHash(int keyHash) {
+            lastKeyHash = keyHash;
         }
 
-        private void setKeyBuffer(KeyBuffer newBuffer) {
-            lastKeyAccessed = newBuffer;
+        int getLastKeyHash() {
+            return lastKeyHash;
         }
 
         private void setKeyValid(boolean keyBufferValid) {
             this.keyBufferValid = keyBufferValid;
         }
 
+        private boolean getKeyValid() {
+            return this.keyBufferValid;
+        }
         @Override
         public void copyState(BasicIteratorState<K, V> other) {
             super.copyState(other);
-            lastKeyAccessed = ((IteratorState<K, V>) other).lastKeyAccessed;
+            lastKeyHash = ((IteratorState<K, V>) other).lastKeyHash;
```

Comment:
I am adding a comment here, because I cannot add it to the line where I see something strange. Just saw it few lines below, as part of the review:

```
@Override
        protected void initAfterRebalance() {
            //TODO - refactor to use OakReadBuffer without deserializing.
            getState().getChunk().readKeyFromEntryIndex(ctx.tempKey, getState().getIndex());
            // Update the state to point to last returned key.
            initState();
}
```

I wonder what for the following lines are part of the ` initAfterRebalance()` for hash? Is the key used later?
```
//TODO - refactor to use OakReadBuffer without deserializing.
getState().getChunk().readKeyFromEntryIndex(ctx.tempKey, getState().getIndex());
```

Reply:
Also thinking about it. `initAfterRebalance()` is absolutely wrong now. Not that we invested thoughts into this implementation. But now when you are keeping the hash of the last key. `initAfterRebalance()` can be implemented 95% correctly. Similar to `advanceState()` just look for the key with the same hash on the new chunk...

Reply:
the function is not implemented, so it can't be wrong :). 
The problem is that splitting a chunk spreads the existing entries between the two new chunks, and the order of keys may change. So , if we just go on scanning, we can either skip some entries or return the same entry twice.

Reply:
That is why I said it is 95% correct and not 100%. Anyway we can refer to it once rebalance is in place.